lib/router/src/cache.rs

```diff
@@ -1,426 +1,402 @@
+                // empty. We know `expirations` is not empty because `vals` is
+                // not empty and capacity does not equal zero.
+                Ok(Async::Ready(None)) => unreachable!("cache expirations cannot be empty"),
+
+                // The cache is at capacity and no values can be removed.
+                Ok(Async::NotReady) => {
+                    return Err(CapacityExhausted {
+                        capacity: self.capacity,
+                    });
+                }
+
+                Err(e) => panic!("Cache.expirations DelayQueue::poll must not fail: {}", e),
             }
         }
 
         Ok(Reserve {
+            expirations: &mut self.expirations,
+            expires: self.expires,
             vals: &mut self.vals,
-            now: &self.now,
         })
     }
 
-    /// Overrides the time source for tests.
-    #[cfg(test)]
-    fn with_clock<M: Now>(self, now: M) -> Cache<K, V, M> {
-        Cache {
-            now,
-            vals: self.vals,
-            capacity: self.capacity,
-            max_idle_age: self.max_idle_age,
+    fn poll_purge(&mut self) -> Poll<(), Error> {
+        while let Some(entry) = try_ready!(self.expirations.poll()) {
+            self.vals.remove(entry.get_ref());
         }
+
+        Ok(Async::Ready(()))
     }
 }
 
-// ===== impl Reserve =====
+// ===== impl PurgeCache =====
 
-impl<'a, K: Hash + Eq + 'a, V: 'a, N: Now + 'a> Reserve<'a, K, V, N> {
-    /// Stores a route in the cache.
-    pub fn store(self, key: K, val: V) {
-        let node = Node::new(val.into(), self.now.now());
-        self.vals.insert(key, node);
-    }
-}
+impl<K: Clone + Eq + Hash, V> Future for PurgeCache<K, V> {
+    type Item = ();
+    type Error = ();
 
-// ===== impl Access =====
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        try_ready!(self.interval.poll().map_err(|e| {
+            panic!("PurgeCache.interval Interval::poll must not fail: {}", e);
+        }));
 
-impl<'a, T: 'a, N: Now + 'a> Deref for Access<'a, T, N> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.node
-    }
-}
+        let lock = match self.cache.upgrade() {
+            Some(lock) => lock,
+            // Failing to upgrade the Weak reference means the cache has been
+            // dropped. We can stop trying to purge values.
+            None => return Ok(Async::Ready(())),
+        };
+        let mut cache = match lock.try_lock() {
+            // If we can lock the cache then do so
+            Ok(lock) => lock,
+
+            // If we were unable to lock the cache, panic if the cause of error
+            // was a poisoned lock
+            Err(TryLockError::Poisoned(e)) => panic!("lock poisoned: {:?}", e),
+
+            // If the lock is not poisoned, it is being held by another
+            // thread. Schedule this thread to be polled in the near future.
+            Err(_) => {
+                futures::task::current().notify();
```

Comment:
If you'd like to use a lock that doesn't block futures, consider using [`tokio::sync::lock::Lock`](https://docs.rs/tokio/0.1.*/tokio/sync/lock/struct.Lock.html) instead. Otherwise, this could cause quite a few spins.

Reply:
I forgot this had merged! :tada: